gonvme_fc.go

```diff
@@ -0,0 +1,359 @@
+
+	var err error
+	// for each initiatior config file
+	for _, init := range initiatorConfig {
+		// make sure the file exists
+		_, err = os.Stat(init)
+		if err != nil {
+			continue
+		}
+
+		// get the contents of the initiator config file
+		// TODO: check if sys call is available for cat command
+		cmd := exec.Command("cat", init)
+
+		out, err := cmd.Output()
+		if err != nil {
+			fmt.Printf("Error gathering initiator names: %v", err)
+		}
+		lines := strings.Split(string(out), "\n")
+
+		for _, line := range lines {
+
+			if line != "" {
+				nqns = append(nqns, line)
+			}
+		}
+	}
+
+	if len(nqns) == 0 {
+		return nqns, err
+	}
+
+	return nqns, nil
+}
+
+// DiscoverNVMeTargets - runs NVMe/FC discovery and returns a list of targets.
+func (nvme *NVMeFC) DiscoverNVMeTargets(targetAddress string, login bool) ([]NVMeTarget, error) {
+	return nvme.discoverNVMeTargets(targetAddress, login)
+}
+
+func (nvme *NVMeFC) discoverNVMeTargets(targetAddress string, login bool) ([]NVMeTarget, error) {
+	// TODO: add injection check on address
+	// nvme discovery is done via nvme cli
+	// nvme discover -t fc -a traddr -w host_traddr
+	// where traddr = nn-<Target_WWNN>:pn-<Target_WWPN> and host_traddr = nn-<Initiator_WWNN>:pn-<Initiator_WWPN>
+
+	var out []byte
+	FCHostsInfo, err := nvme.getFCHostInfo()
+	if err != nil {
+		fmt.Printf("Error gathering NVMe/FC Hosts on the host side: %v", err)
+		return []NVMeTarget{}, nil
+	}
+
+	targets := make([]NVMeTarget, 0)
+	for _, FCHostInfo := range FCHostsInfo {
+
+		// host_traddr = nn-<Initiator_WWNN>:pn-<Initiator_WWPN>
+		initiatorAddress := strings.Replace(fmt.Sprintf("nn-%s:pn-%s", FCHostInfo.NodeName, FCHostInfo.PortName), "\n", "", -1)
+		exe := nvme.buildNVMeCommand([]string{NVMeCommand, "discover", "-t", "fc", "-a", targetAddress, "-w", initiatorAddress})
+		cmd := exec.Command(exe[0], exe[1:]...)
+
+		out, err = cmd.Output()
+		if err != nil {
+			continue
+		}
+
+		nvmeTarget := NVMeTarget{}
+		entryCount := 0
+		skipIteration := false
+
+		for _, line := range strings.Split(string(out), "\n") {
+
+			// Output should look like:
+
+			// Discovery Log Number of Records 2, Generation counter 2
+			// =====Discovery Log Entry 0======
+			// trtype:  fc
+			// adrfam:  fibre-channel
+			// subtype: nvme subsystem
+			// treq:    not specified
+			// portid:  0
+			// trsvcid: none
+			// subnqn:  nqn.1111-11.com.dell:powerstore:00:a1a1a1a111a1111a111a
+			// traddr:  nn-0x11aaa111a1111a11:aa-0x11aaa11111111a11
+			//
+			// =====Discovery Log Entry 1======
+			// trtype:  tcp
+			// adrfam:  ipv4
+			// subtype: nvme subsystem
+			// treq:    not specified
+			// portid:  2304
+			// trsvcid: 4420
+			// subnqn:  nqn.1111-11.com.dell:powerstore:00:a1a1a1a111a1111a111a
+			// traddr:  1.1.1.1
+			// sectype: none
+
+			tokens := strings.Fields(line)
+			if len(tokens) < 2 {
+				continue
+			}
+			key := tokens[0]
+			value := strings.Join(tokens[1:], "")
+			switch key {
+
+			case "=====Discovery":
```

Comment:
please fix indentation.

Reply:
The indentation changes back to this on running the gofmt command

Reply:
ok